File: contest/free_contest_149/count/main.py
import sys
import math

def cnt(list_num = [], n = 0):
    lcm = 1
    for num in list_num:
        lcm = math.lcm(lcm, num)
    return int(n / lcm)

for line in sys.stdin:
    s = line.split(" ")
    n = int(s[0])
    p = int(s[1])
    q = int(s[2])
    r = int(s[3])
    ans = cnt([p, q], n) + cnt([q, r], n) + cnt([p, r], n) - 3 * cnt([p, q, r], n)
    print(ans)

File: contest/free_contest_149/count/main.cpp
#include <bits/stdc++.h>
#define ll long long
#define endl '\n'

using namespace std;

int32_t main() {
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    ll n, p, q, r;
    auto cnt = [&](ll x, ll y) -> ll {
        return n / lcm(x, y);
    };
    while(cin >> n >> p >> q >> r) {
        ll ans = cnt(p, q) + cnt(q, r) + cnt(p, r);
        ll lcm_pq = lcm(p, q);
        ll gcd_pqr = gcd(lcm_pq, r);
        lcm_pq /= gcd_pqr;
        r /= gcd_pqr;
        if (lcm_pq * r <= (ll)1e18 / gcd_pqr) {
            ans -= n / (lcm_pq * r * gcd_pqr) * 3;
        }
        cout << ans << endl;
    }
    return 0;
}